File: src/manifest.ts
/**
 * Manifest fetcher + cache.
 *
 * Each app publishes a manifest at `<app>/api/v2/manifest` describing
 * the full feature.action surface (auth modes + Zod-derived JSON schemas
 * for input/output). The CLI caches manifests under
 * `~/.mm-cli/manifests/<slug>.json` so subsequent commands can
 * pre-validate input + render --help text without a network round trip.
 *
 * Cache invalidation: `mm manifest <app> --refresh` or 24h TTL.
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync, statSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';
import { resolveApp } from './apps';

const CACHE_DIR = join(homedir(), '.mm-cli', 'manifests');
const CACHE_TTL_MS = 24 * 60 * 60 * 1000; // 24h

export type AuthMode = 'session' | 'hub' | 'either' | 'public' | 'install';

export interface ManifestAction {
	auth: AuthMode;
	description?: string;
	input: Record<string, unknown> | 'Response';
	output: Record<string, unknown> | 'Response';
}

export interface AppManifest {
	appSlug: string;
	version: string;
	features: Record<string, Record<string, ManifestAction>>;
}

function ensureDir() {
	if (!existsSync(CACHE_DIR)) {
		mkdirSync(CACHE_DIR, { recursive: true });
	}
}

function cachePath(slug: string): string {
	return join(CACHE_DIR, `${slug}.json`);
}

function isCacheFresh(path: string): boolean {
	try {
		const age = Date.now() - statSync(path).mtimeMs;
		return age < CACHE_TTL_MS;
	} catch {
		return false;
	}
}

/**
 * Fetch a manifest from the app's `/api/v2/manifest` endpoint.
 * No auth — manifests are public surface descriptions.
 */
export async function fetchManifest(slug: string): Promise<AppManifest> {
	const app = resolveApp(slug);
	const url = `${app.url}/api/v2/manifest`;
	const res = await fetch(url, { headers: { accept: 'application/json' } });
	if (!res.ok) {
		throw new Error(`Manifest fetch ${url} failed: HTTP ${res.status}`);
	}
	const manifest = (await res.json()) as AppManifest;
	if (!manifest.appSlug || !manifest.features) {
		throw new Error(`Manifest from ${url} is malformed: missing appSlug/features`);
	}
	return manifest;
}

/**
 * Load a manifest from cache (if fresh) or fetch + cache.
 * Pass `{refresh: true}` to bypass cache.
 */
export async function loadManifest(
	slug: string,
	opts: { refresh?: boolean } = {}
): Promise<AppManifest> {
	ensureDir();
	const path = cachePath(slug);

	if (!opts.refresh && existsSync(path) && isCacheFresh(path)) {
		try {
			return JSON.parse(readFileSync(path, 'utf-8')) as AppManifest;
		} catch {
			// Corrupt cache — fall through to refetch.
		}
	}

	const manifest = await fetchManifest(slug);
	writeFileSync(path, JSON.stringify(manifest, null, 2));
	return manifest;
}

export function clearManifestCache(slug?: string) {
	ensureDir();
	if (slug) {
		const p = cachePath(slug);
		if (existsSync(p)) {
			require('node:fs').unlinkSync(p);
		}
	} else {
		// Clear all
		for (const f of require('node:fs').readdirSync(CACHE_DIR)) {
			require('node:fs').unlinkSync(join(CACHE_DIR, f));
		}
	}
}

/**
 * Total action count across all features in a manifest. Useful for
 * `mm manifest` summary output.
 */
export function actionCount(m: AppManifest): number {
	return Object.values(m.features).reduce((n, actions) => n + Object.keys(actions).length, 0);
}

/**
 * Resolve a `feature.action` string against a manifest. Throws with a
 * helpful message if either part is missing.
 */
export function resolveAction(
	manifest: AppManifest,
	featureAction: string
): { feature: string; action: string; def: ManifestAction } {
	const dot = featureAction.indexOf('.');
	if (dot < 0) {
		throw new Error(
			`feature.action must be 'feature.action' format, got: '${featureAction}'`
		);
	}
	const feature = featureAction.slice(0, dot);
	const action = featureAction.slice(dot + 1);
	const featureMap = manifest.features[feature];
	if (!featureMap) {
		const known = Object.keys(manifest.features).slice(0, 10).join(', ');
		throw new Error(
			`Unknown feature '${feature}' in ${manifest.appSlug}. ` +
				`Known: ${known}${Object.keys(manifest.features).length > 10 ? '…' : ''}`
		);
	}
	const def = featureMap[action];
	if (!def) {
		const known = Object.keys(featureMap).slice(0, 10).join(', ');
		throw new Error(
			`Unknown action '${action}' on ${manifest.appSlug}.${feature}. ` +
				`Known: ${known}${Object.keys(featureMap).length > 10 ? '…' : ''}`
		);
	}
	return { feature, action, def };
}

File: internal/manifest/manifest.go
// Package manifest fetches + caches `<app>/api/v2/manifest`.
// Mirrors src/manifest.ts: ~/.mm-cli/manifests/<slug>.json, 24h TTL.
package manifest

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"mm-cli/internal/apps"
)

const cacheTTL = 24 * time.Hour

type Action struct {
	Auth        string `json:"auth"`
	Description string `json:"description,omitempty"`
	// Input/Output are arbitrary JSON Schema-ish payloads; we just preserve them.
	Input  json.RawMessage `json:"input"`
	Output json.RawMessage `json:"output"`
}

type Manifest struct {
	AppSlug  string                       `json:"appSlug"`
	Version  string                       `json:"version"`
	Features map[string]map[string]Action `json:"features"`
}

func cacheDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	d := filepath.Join(home, ".mm-cli", "manifests")
	return d, os.MkdirAll(d, 0o755)
}

func cachePath(slug string) (string, error) {
	d, err := cacheDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(d, slug+".json"), nil
}

// Load returns the manifest for `slug`, from cache if fresh (24h) or by
// fetching `<app>/api/v2/manifest`. Pass `refresh=true` to bypass cache.
func Load(ctx context.Context, slug string, refresh bool) (*Manifest, error) {
	path, err := cachePath(slug)
	if err != nil {
		return nil, err
	}
	if !refresh {
		if fi, err := os.Stat(path); err == nil && time.Since(fi.ModTime()) < cacheTTL {
			data, err := os.ReadFile(path)
			if err == nil {
				var m Manifest
				if json.Unmarshal(data, &m) == nil {
					return &m, nil
				}
			}
		}
	}
	m, err := Fetch(ctx, slug)
	if err != nil {
		return nil, err
	}
	data, _ := json.MarshalIndent(m, "", "  ")
	_ = os.WriteFile(path, data, 0o644)
	return m, nil
}

func Fetch(ctx context.Context, slug string) (*Manifest, error) {
	app, err := apps.Resolve(slug)
	if err != nil {
		return nil, err
	}
	url := app.URL + "/api/v2/manifest"
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	req.Header.Set("Accept", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("manifest fetch %s: %w", url, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode/100 != 2 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("manifest fetch %s failed: HTTP %d: %s", url, resp.StatusCode, string(body))
	}
	var m Manifest
	if err := json.NewDecoder(resp.Body).Decode(&m); err != nil {
		return nil, fmt.Errorf("manifest %s: invalid JSON: %w", url, err)
	}
	if m.AppSlug == "" || m.Features == nil {
		return nil, fmt.Errorf("manifest from %s is malformed: missing appSlug/features", url)
	}
	return &m, nil
}

// ActionCount returns the total feature.action count.
func (m *Manifest) ActionCount() int {
	n := 0
	for _, actions := range m.Features {
		n += len(actions)
	}
	return n
}

// ResolveAction splits "feature.action" and looks it up, returning a clean
// error if either part is missing.
func (m *Manifest) ResolveAction(featureAction string) (feature, action string, def Action, err error) {
	dot := strings.IndexByte(featureAction, '.')
	if dot < 0 {
		return "", "", Action{}, fmt.Errorf("feature.action must be 'feature.action' format, got: '%s'", featureAction)
	}
	feature = featureAction[:dot]
	action = featureAction[dot+1:]
	fm, ok := m.Features[feature]
	if !ok {
		known := truncStringList(keys(m.Features), 10)
		return "", "", Action{}, fmt.Errorf("Unknown feature '%s' in %s. Known: %s", feature, m.AppSlug, known)
	}
	d, ok := fm[action]
	if !ok {
		known := truncStringList(keys(fm), 10)
		return "", "", Action{}, fmt.Errorf("Unknown action '%s' on %s.%s. Known: %s", action, m.AppSlug, feature, known)
	}
	return feature, action, d, nil
}

func keys[V any](m map[string]V) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	return out
}

func truncStringList(xs []string, n int) string {
	if len(xs) <= n {
		return strings.Join(xs, ", ")
	}
	return strings.Join(xs[:n], ", ") + "…"
}
